{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6be17027-2256-47e2-82ad-6173fdf45a62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c99c6c3-e03d-463e-b865-dcf8050c04ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import\n",
    "import integrative_transcriptomics_viewer as itv\n",
    "from integrative_transcriptomics_viewer.convenience import Configuration\n",
    "from integrative_transcriptomics_viewer.export import save\n",
    "from integrative_transcriptomics_viewer.classification import BAMtagClassification\n",
    "from integrative_transcriptomics_viewer.annotation_matching import TaggedBAMAnnotationMatching\n",
    "from intervaltree import Interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15e673f8-1f43-4aba-a89a-8fb7e4a9f9ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference files \n",
    "human_genome_path = \"genome.fa\"\n",
    "human_gtf_path = \"genes.gtf\"\n",
    "human_bed_gencode = \"genes.bed.gz\"\n",
    "\n",
    "human = Configuration(genome_fasta = human_genome_path, gtf_annotation = human_gtf_path, bed_annotation = [str(human_bed_gencode)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb218336-b613-4f98-a10b-088a4905ffbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input BAM file\n",
    "all_bams_dict = {\"Short\":\"SR.PBMC.sort.bam\",\n",
    "                \"BenchDrop-seq\":\"LR.PBMC.genome.sort.bam\",\n",
    "                \"Bulk\":\"out.Aligned.sortedByCoord.out.bam\"} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa06127-7f5e-4952-9145-7b9771de5a11",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"CORO1B\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a4d26dd-3666-4554-910d-254e96b68ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# View in notebook\n",
    "\n",
    "human.plot_feature(gene, \n",
    "    bams_dict = all_bams_dict, \n",
    "    with_reads = False,\n",
    "    tighter_track = True,\n",
    "    add_track_label = gene,\n",
    "    add_coverage_label = True,\n",
    "    vertical_layout_reads = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9469cf43-539c-483f-96ec-85d28f5cb0d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save plot\n",
    "\n",
    "plot = human.plot_feature(gene, \n",
    "    bams_dict = all_bams_dict, \n",
    "    with_reads = False,\n",
    "    tighter_track = True,\n",
    "    add_track_label = gene,\n",
    "    add_coverage_label = True,\n",
    "    vertical_layout_reads = True)\n",
    "\n",
    "save(plot, \"Fig3.svg\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
